[REVIEW] Break out benchmark notebook utils and expand (v2)

JohnZed opened the pull request:
This has a lot of rough edges - few docstrings, no tests, and some functions are too big. But I was curious to get feedback on the direction.

I started by taking the classes defined at the top of the benchmarks notebook and adding them to "cuml.benchmarks". I expanded them to meet a couple of feature requests:
 * Support for command-line benchmarks in addition to notebooks
 * Flexible datasets (e.g. use real-world data like Higgs in addition to make_regression)
 * Dump results as CSV or table
 * Output wall clock time (in addition to comparison with sklearn)
 * Support cuml-only benchmarks (e.g. holt-winters may not have an sklearn equivalent, or we may want to do our own scaling tests to larger sizes and not wait for the sklearn version to run)
  * Make it shorter to write wrappers for new algos

I added two example scripts:
 * bench_all - potentially useful utility that's basically a command-line version of the notebook
 * bench_kmeans - trivial example, won't check it in, but simpler than bench_all

Supporting flexible dataset backends and sizes is the gnarliest part for the user interface. E.g. for a real-world dataset, you often just want "give me the default size" or "default num features but only 10k rows," while for synthetics we have to specify feature counts and row sizes always. Hence the weird stuff around default sizes... probably there is some nicer approach possible. Also, by default we pretend every dataset is for supervised learning then allow algorithms to throw away the `y` if they don't need it. That seemed to simplify the interface a bit. I do think it will be easy to add all of the datasets from gbm-benchmarks trivially after this. As noted, I think it will also be easy to move to using cudf for datasets, but that can be a future change.

I didn't produce a demo version of the notebook with this code, but I think it'll be pretty easy. I definitely still want to maintain the notebook and graphing support, since that's the most visual approach we have.

cjnolet commented on python/cuml/benchmark/bench_all.py:
```diff
@@ -0,0 +1,113 @@
+
+def get_all_algos(load_data):
+    return [
+        utils.AlgoComparisonWrapper(
```
Mostly thinking out loud here. 

I like the simplicity behind this. I'm wondering if we should define these wrapper instances as globals somewhere, and whether we should specify different values/ranges of the `shared_args` to try, similar to `pytest.mark.parameter`

For example, it would be really neat if we had a global kmeans definition that we could just reuse everywhere else, like this:

```
kmeans_wrapper = utils.AlgoComparisonWrapper(
    sklearn.cluster.KMeans,
    cuml.cluster.KMeans,
    shared_args=dict(init = ["random", "kmeans||"], 
                                 n_clusters=np.logspace(2.0, 3.0, num=5), 
                                 max_iter=300]),
    name='kmeans')
) 
...
```

While this is being redesigned, I like the idea of abstracting the dataset away from the algorithm being benchmarked so that the benchmark being run can take care of loading the data, converting to the proper format, and running it through an `AlgoComparisonWrapper`.

Perhaps there could be a `Dataset` class that can be constructed w/ an array of data, which contains the relevant conversion functions. 

```
class Dataset(object):
    def load_data(n_rows, n_dims, out_type = 'cudf'):
        pass

class HiggsDataset(Dataset):
    def __init__(fname = "..."):
        # read higgs dataset into some internal structure

    def load_data(n_rows, n_dims, out_type = 'cudf'):
        return X or X,y depending on the type of dataset. 
```        
    

It would be neat if the interactions between objects could be something like below:

```
datasets = [HiggsDataset(fname=".."), BlobsDataset(n_clusters=5)...]

br = BenchmarkRunner(
    benchmarks = [utils.SpeedupBenchmark],
    data_types = ['cudf', 'numpy'],
    verbose = true,
    bench_rows = ...,
    bench_dims = ...
    n_trials = 3
)

results = BenchmarkResults()
for d in datasets:
    for a in algowrappers:
        results.combine(br.run(a, d,))

print(results.to_dataframe())
results.to_csv("results.csv")

BenchmarkChart(results).chart_sizes_by_dims().show()
```

This way, the benchmark runner is responsible for orchestrating the interactions between instances of `Benchmark`, `AlgoComparisonWrapper`, and `Dataset`

I also think the chart drawing should get separated from the `BenchmarkRunner` class. Perhaps the benchmark runner could produce an instance of  a `BenchmarkResults` class, which would know how store/load itself to/from disk (as csv, pickle, whatever). Default behavior would be for it to just to store a dict/dataframe in memory w/ the results of the benchmark:

```
class DataPoint(object):
    def __init__(self,n_rows, n_dims, algo_name, shared_args, dict(time_sk=..., time_cuml=..., speedup=...)

class BenchmarkResults(object):
    def combine(other_benchmark_results)
    def add_result(data_point):
        # Adds this entry to whatever data struct it is using internally to store the result of a single benchmark run. 

    def to_csv(self)
    def to_dataframe(self)
    def __setstate__(self)
    def __getstate__(self)
```

Charting could be its own beast, which could take a `BenchmarkResults` instance and draw different types of charts:

```
class BenchmarkChart(object):
    def __init__(self, benchmark_results)

    # The chart currently in the code
    def chart_sizes_by_dims(algo_name, shared_args, min_n_rows=..., max_n_rows=..., min_n_dims=..., max_n_dims=...)

    # Compares the shared args for a single algorithm
    def chart_compare_shared_args(algo_name, min_n_rows=..., max_n_rows=..., min_n_dims=...)

    # Compares algo times across several algorithms (algo_names=None compares all algos in result set), maybe aggregating across different shared args results. 
    def chart_compare_algo_times(algo_names = None, min_n_rows=..., max_n_rows=..., min_n_dims=...)
```

What do you think?

cjnolet replied:
> are you thinking that a dataset should include the nrows and nfeatures? 

Sure. Even with fixed-sized datasets we could sample uniformly to get any number of rows/features we need for a particular volume. Or perhaps set reasonable defaults that automatically use the default size of the dataset. 

I think the pytest parametrization is a great idea!

cjnolet commented:
I wanted to leave a note here to check out @pentschev's new benchmark lib, which uses Pytest parameter markings, very similar to what I've proposed on comments in this PR. 

I think we should generalize the data loading and charting API, but we should also strive to use a shared tool across RAPIDS where possible.

pentschev commented:
Thanks @cjnolet for pointing it here. I was also talking to @rlratzel earlier today, that's also something that CI might be interested in. So hopefully we can have a single tool for RAPIDS benchmarking everywhere.

JohnZed commented:
Hi, @pentschev , yes, we've been talking with Rick too and are interested in figuring out the best way to support pytest as one frontend. I think one conclusion we came too is that each library has some core needs (like the bench_algos and bench_data stuff in this PR) and then there are a variety of runners and wrappers that can be built on top of those cores. In particular, regression testing microbenchmarks, deep dive parameter sweep benchmarks, and easy-to-use notebook speedup demos all share a core but may have slightly different needs.

In the latest version of this PR, I include a pytest-based frontend based on pytest-benchmark. It was very easy to support, so I think that supporting pybench would be straighforward as well.

JohnZed commented:
@cjnolet I finally got this updated, though it's still WIP. The new rev:
 * Removes some of the old notebook-specific code like charting... we can add that into the notebook frontend or another module directly but I think it was just cluttering things here
 * Moves algo default definitions to a shared location per suggestion
 * Cleans up dataset generation and separates it more cleanly from algo running
 * Adds a demo frontend based on pytest-benchmark.

@rlratzel is looking into how hard it will be to integrate ASV automatically with pytest-benchmark or something similar. Failing that, the bench_all.py script could be integrated pretty easily with Rick's CSV dumper, we'd just need a trivial bash script to run it with the desired args and to integrate with his simple API.

cjnolet commented on python/cuml/benchmark/bench_runners.py:
```diff
@@ -0,0 +1,143 @@
+                y_pred_cuml = cuml_model.transform(X_test)
+            cuml_accuracy = algo_pair.accuracy_function(y_test, np.asarray(y_pred_cuml))
+        else:
+            cuml_accuracy = 0.0
```
Would it make sense to just compare the outputs directly in the case no accuracy function has been provided? I'm thinking of NearestNeighbors here. Or maybe to compare the results with some predefined tolerance?

JohnZed commented:
Ok, this is ready to go at last. Cleaned up formatting, added tests, etc.
This includes a CI-oriented tool that requires `asvdb` from @rlratzel , which is not installed by default yet but will be in the near future on CI machines.
The `run_benchmarks` tool now includes examples.
It should be pretty isolated, but if you can give a quick look @cjnolet or @rlratzel , that would be much appreciated.

JohnZed commented:
rerun tests

rlratzel commented:
FYI, for now you can install `asvdb` with conda like this:
```
conda install -c rlratzel asvdb
```

cjnolet commented:
I just realized I had some open questions here that remain unanswered.

JohnZed commented on python/cuml/benchmark/bench_algos.py:
```diff
@@ -0,0 +1,200 @@
+            name='MBSGDClassifier',
+            accepts_labels=True,
+            accuracy_function=cuml.metrics.accuracy_score),
+    ]
```
I added TSVD. Will add random projection shortly.

JohnZed replied:
Added random projection.

cjnolet commented:
Sounds good. We should add UMAP supervised learning as well (just use a
dataset w/ labels)

On Thu, Sep 19, 2019 at 6:33 PM John Zedlewski <notifications@github.com>
wrote:

> *@JohnZed* commented on this pull request.
> ------------------------------
>
> In python/cuml/benchmark/bench_algos.py
> <https://github.com/rapidsai/cuml/pull/855#discussion_r326410097>:
>
> > +            accuracy_function=metrics.r2_score),
> +        AlgorithmPair(
> +            umap.UMAP,
> +            cuml.manifold.UMAP,
> +            shared_args=dict(n_neighbors=5, n_epochs=500),
> +            name='UMAP',
> +            accepts_labels=False,
> +            accuracy_function=cuml.metrics.trustworthiness),
> +        AlgorithmPair(
> +            None,
> +            cuml.linear_model.MBSGDClassifier,
> +            shared_args={},
> +            name='MBSGDClassifier',
> +            accepts_labels=True,
> +            accuracy_function=cuml.metrics.accuracy_score),
> +    ]
>
> I added TSVD. Will add random projection shortly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rapidsai/cuml/pull/855?email_source=notifications&email_token=AAJPKYC7QSMLXRP5N566ACTQKP45BA5CNFSM4IEI4OQ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFLHB7Q#discussion_r326410097>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJPKYE7CLHM3HYO5T5QS3LQKP45BANCNFSM4IEI4OQQ>
> .
>

JohnZed commented on python/cuml/benchmark/bench_kmeans.py:
```diff
@@ -0,0 +1,27 @@
+    shared_args=dict(init='random',
+                     n_clusters=8,
+                     max_iter=300),
+    accepts_labels=False,
```
Specifically I'm using this to override defaults. If the defaults are good, we can just stick with them automatically.
